pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1164,40 +1164,40 @@ func (vs *VSphere) DisksAreAttached(nodeVolumes map[k8stypes.NodeName][]string)
 func (vs *VSphere) CreateVolume(volumeOptions *vclib.VolumeOptions) (canonicalVolumePath string, err error) {
 	klog.V(1).Infof("Starting to create a vSphere volume with volumeOptions: %+v", volumeOptions)
 	createVolumeInternal := func(volumeOptions *vclib.VolumeOptions) (canonicalVolumePath string, err error) {
-		var datastore string
+		var datastoreInfo *vclib.DatastoreInfo
 		var dsList []*vclib.DatastoreInfo
-		// If datastore not specified, then use default datastore
-		if volumeOptions.Datastore == "" {
-			datastore = vs.cfg.Workspace.DefaultDatastore
-		} else {
-			datastore = volumeOptions.Datastore
-		}
-		datastore = strings.TrimSpace(datastore)
+
 		// Create context
 		ctx, cancel := context.WithCancel(context.Background())
 		defer cancel()
 		vsi, err := vs.getVSphereInstanceForServer(vs.cfg.Workspace.VCenterIP, ctx)
 		if err != nil {
 			return "", err
 		}
-		dc, err := vclib.GetDatacenter(ctx, vsi.conn, vs.cfg.Workspace.Datacenter)
+		// If datastore not specified, then use default datastore
+		datastoreName := strings.TrimSpace(volumeOptions.Datastore)
+		if datastoreName == "" {
+			datastoreName = strings.TrimSpace(vs.cfg.Workspace.DefaultDatastore)
+		}
+		datastoreInfo, err = vs.FindDatastore(ctx, datastoreName)
```

Comment:
How would volume creation work if allowedTopologies is specifying zone-1 and this code is selecting a datastore in zone-2?

Reply:
Changed this to consider all matching datastores (across datacenters) and allow the other inputs like zone or storage profile to pick out one of these datastores that also match the zone or storage profile.

Reply:
Changed this to consider all matching datastores (across datacenters) and allow the other inputs like zone or storage profile to pick out one of these datastores that also match the zone or storage profile.